### PR TITLE
HAProxy PROXY protocol support.

### PR DESCRIPTION
**Pull request by btimby**
This is only a sketch, I will test it tomorrow. I can add some unit tests to this project if you like, I also have unit tests that will exercise this code in my own project.

Please give feedback and I can address it tomorrow before I start testing.

### CODE REVIEW

**Comment by btimby** [comment]
OK, I am happy with the docs. Code is way easier to write than docs.

File: lib/server.js [hunk 1]
```diff
@@ -44,85 +142,22 @@ module.exports = function (server, options) {
...
-                    manage(c, address);
-                }());
-                break;
-            default:
-                return invalidHeader("Invalid Header Type", {type: type, length: length});
-            }
+    server.on('connection', function(c) {
+        c.on('readable', function(c) {
+            parser(c, manage);
             server.emit('wrappedConnection', c);
```
Comment (B3rn475) [inline lib/server.js]: This is not equivalent.
The `wrappedConnection` event was triggered at the end of the header parsing, which can require more than one `readable` event (the connection can be slow and not all the headers is returned).

File: lib/server.js [hunk 2]
```diff
@@ -44,85 +142,22 @@ module.exports = function (server, options) {
...
-                    address = ipaddr.parse(address).toString();
-                    manage(c, address);
-                }());
-                break;
-            default:
-                return invalidHeader("Invalid Header Type", {type: type, length: length});
-            }
+    server.on('connection', function(c) {
+        c.on('readable', function(c) {
+            parser(c, manage);
```
Comment (B3rn475) [inline lib/server.js]: What is `manage`?

**Comment by B3rn475** [comment]
I did a refactoring to facilitate the drop-in of new header formats and testing.

I think it should be easier now to integrate it.

File: lib/server/header/haproxy.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
+function parseHeader(header) {
+    var parts = header.split(' ')[2];
+    if (parts[0] !== 'PROXY') {
+        throw new Error('PROXY line missing, is haproxy is present?');
+    }
+    return parts[2];
+}
+
+function createParser(socket, connected, headerError) {
+    var buffer,
```
Comment (B3rn475) [inline lib/server/header/haproxy.js]: You should initialize the `buffer` to to an empty **string**

File: lib/server/header/haproxy.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        char;
```
Comment (B3rn475) [inline lib/server/header/haproxy.js]: The `char` variable is not shared between executions, it should be in the function.

File: lib/server/header/haproxy.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+    return function () {
+        while (true) {
+            if ((char = socket.read(1)) === null) {
+                return;
+            }
+            if (char === '\n') {
```
Comment (B3rn475) [inline lib/server/header/haproxy.js]: `read(1)` returns a **buffer**, you need to convert it to a **string** with `.toString('ascii')` before using it.

**Comment by B3rn475** [comment]
I think after the comments it should be ok.

Can you still add some tests?

**Comment by btimby** [comment]
I addressed your comments, I added tests and fixed my additional mistakes. I think it is good to go now.

**Pull request merged**